Support multiple instance configurations

jeremyh opened the pull request:
### Reason for this pull request

* We'd like to able to use mulitple datacube environments without manually managing separate config files.
* GA would prefer to only deploy and manage one `dea` module on NCI, instead of three (`-prod`, `-staging`, `dev`)
* It would be nice if users could more "effortlessly" use their own custom dea instances (eg. `WOfSCube`) without the pain of interference with existing prod config.
* The current config setup makes it easier than it should be to accidentally use the wrong instance. We'd to allow code to be explicit ("do this in prod" / "do this in staging") without users manually writing and managing their own separate config files everywhere.

### Proposed changes

This extends the LocalConfig classes to allow the user to specify multiple datacube environments in one config file

eg.

    [user]
    default_environment: dev

    [dev]
    db_hostname: localhost
    db_database: dev

    [staging]
    db_database: staging

Then 'dev' is used by default, but staging can be specified during any setup:

eg.

```
$ datacube -E staging system check
Version:       1.5.1+176.gaf2c2fee
Config files:  /home/jez/.datacube.conf
Host:          localhost:5432
Database:      staging
User:          jez
Environment:   staging
$
```

or

```
with Datacube(env='staging') as dc:
    ...
```

Note that this change is backwards compatible with all existing user configs: if not specified, the default environment is `datacube`, which is the config section name used on previous config files.

jeremyh committed:
Simplify Datacube constructor logic

jeremyh committed:
Allow use of multiple config sections for separate environments

DCDT-290

jeremyh committed:
Allow environment to be specified by users on cli

DCDT-290

jeremyh committed:
Allow user to set their default environment

They could add a custom one in ~/.datacube.conf or
other.

jeremyh committed:
Add clarifying comments to environment config options

jeremyh committed:
Environment variable fallbacks should work for manually created configs

This gives LocalConfig.find() and LocalConfig() the same defaults.

The Code is also more easily understandable with the chain
of 'or' fallbacks in one place.

jeremyh committed:
Allow config to specify a default driver

jeremyh committed:
Show the effective environment in `system check` command

jeremyh committed:
Simplify `system check` field printing

jeremyh committed:
Update docs for multi-environment usage

jeremyh committed:
Support env option when no config is specified

jeremyh committed:
Add multi-environments to release notes

Kirill888 reviewed (dismissed):
Good UX improvement. I'm all for it.

jeremyh committed:
Remove legacy 'locations' section from test config

jeremyh committed:
Add tests for multiple envionments

jeremyh committed:
Fix lazy connection handling in DriverManager

validate_connection=False is supposed to stop any connections to be
made on startup, but the DriverManager was creating them.

jeremyh committed:
Support py2's config parsing limitations

Python2 requires the section header to start at the beginning
of the line (no whitespace), this was causing the 2.7 test to fail.

jeremyh commented:
@Kirill888 I added a test and fixed a connection issue since your previous review. Ready to merge if you're ok with it.
